In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Ratio Imputation**



In ratio imputation we calculate the ratio of missing values and impute them based on the threshold set. Ratio can be calculated using simple formula : 

<html> <p style="color:red"> Ratio of missing values= (Number of missing values/Total number of Observations) * 100 </p> </html>


In [ ]:
#importing the libraries

import pandas as pd
import numpy as np


In [ ]:
#reading the file
df = pd.read_csv('video_games_sales.csv')

#first 5 rows of the data
df.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  
0          51.0          8       322.0  Nintendo      E  
1           NaN        NaN         NaN       NaN    NaN  
2          73.0        8.3       709.0  Nintendo      E  
3          73.0          8       192.0  Nintendo      E  
4           NaN        NaN         NaN       NaN    NaN

In [ ]:
# To add teh NAN values to any column in the dataset (doing this to provide better example of Ratio Imputation in this case)
df.loc[100:13000,'Rating']=np.NAN
df.loc[100:7000,'User_Count']=np.NAN

In [ ]:
#shape of the data

df.shape

(16719, 16)

We have 16719 observations and 16 variables or columns. Next, we will check the number of missing values in each column-

calculate the missing values in each column

In [ ]:
#number of missing values in each variable

df.isnull().sum()

Name                   2
Platform               0
Year_of_Release      269
Genre                  2
Publisher             54
NA_Sales               0
EU_Sales               0
JP_Sales               0
Other_Sales            0
Global_Sales           0
Critic_Score        8582
Critic_Count        8582
User_Score          6704
User_Count         13269
Developer           6623
Rating             15082
dtype: int64

Here you can see that Rating and User_Count seem to have a very high number of missing observations. The variables Critic_Score ,Critic_Count,User_score and developer also have missing values, as you can see, but they’re pretty negligible compared to rating and user_count.

calculate the Ratio (percentage of the missing values)

In [ ]:
len(df)

16719

In [ ]:
# percentage of missing values in each variable

df.isnull().sum()/len(df)*100

Name                0.011962
Platform            0.000000
Year_of_Release     1.608948
Genre               0.011962
Publisher           0.322986
NA_Sales            0.000000
EU_Sales            0.000000
JP_Sales            0.000000
Other_Sales         0.000000
Global_Sales        0.000000
Critic_Score       51.330821
Critic_Count       51.330821
User_Score         40.098092
User_Count         79.364795
Developer          39.613613
Rating             90.208745
dtype: float64

Once we have the missing value ratio on the variables, what’s the next step? Now we can decide a threshold, let’s say 70% and you can use this threshold and drop all the variables which have a missing value ratio more than this threshold. So when we look at the ratio "the Rating" variable has a missing value of 90% which is way more than our threshold of 70%. Hence we can go ahead and delete the variable (column). Generally we can delete variables having missing value ratio which is more than 60% or 70% 

**Keep in Mind** There is no hard and fast rule for setting the threshold, it will vary from problem to problem

**Time to Think** Once you drop the variables, which are missing values more than the threshold, how do you think we should deal with the remaining variables, which still have missing values in them? Should we just go ahead and remove all of them?


As rating and user count has missing value above 70% which is quite a lot. If a variable has around 40-50% missing value it is usually imputed using mean, median, mode

In [ ]:
# saving missing values in a variable
a = df.isnull().sum()/len(df)*100

# saving column names in a variable
variables = df.columns

In [ ]:
df.columns

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales', 'Critic_Score',
       'Critic_Count', 'User_Score', 'User_Count', 'Developer', 'Rating'],
      dtype='object')

In [ ]:
a

Name                0.011962
Platform            0.000000
Year_of_Release     1.608948
Genre               0.011962
Publisher           0.322986
NA_Sales            0.000000
EU_Sales            0.000000
JP_Sales            0.000000
Other_Sales         0.000000
Global_Sales        0.000000
Critic_Score       51.330821
Critic_Count       51.330821
User_Score         40.098092
User_Count         79.364795
Developer          39.613613
Rating             90.208745
dtype: float64

In [ ]:
variables

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales', 'Critic_Score',
       'Critic_Count', 'User_Score', 'User_Count', 'Developer', 'Rating'],
      dtype='object')

Using the For Loop to decrease the dimensionality of the dataset by removing the columns having missing value above the set threshold


In [ ]:
# new variable to store variables having missing values less than a threshold

variable = [ ]

for i in range(df.columns.shape[0]):
    if a[i]<=70: #setting the threshold as 70%
        variable.append(variables[i])

In [ ]:
variable

['Name',
 'Platform',
 'Year_of_Release',
 'Genre',
 'Publisher',
 'NA_Sales',
 'EU_Sales',
 'JP_Sales',
 'Other_Sales',
 'Global_Sales',
 'Critic_Score',
 'Critic_Count',
 'User_Score',
 'Developer']

In [ ]:
# creating a new dataframe using the above variables

new_df = df[variable]
new_df.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score Developer  
0          51.0          8  Nintendo  
1           NaN        NaN       NaN  
2          73.0        8.3  Nintendo  
3          73.0          8  Nintendo  
4           NaN        NaN       NaN

In [ ]:
# percentage of missing values in each variable of new data

new_df.isnull().sum()/len(new_df)*100

Name                0.011962
Platform            0.000000
Year_of_Release     1.608948
Genre               0.011962
Publisher           0.322986
NA_Sales            0.000000
EU_Sales            0.000000
JP_Sales            0.000000
Other_Sales         0.000000
Global_Sales        0.000000
Critic_Score       51.330821
Critic_Count       51.330821
User_Score         40.098092
Developer          39.613613
dtype: float64

In [ ]:
# shape of new and original data

new_df.shape, df.shape


((16719, 14), (16719, 16))

the number of observations is the same, but there are two fewer variables in the new dataset.